File: static/js/performance.js
async function perfLoad(view) {
  spin('perfContent');
  try {
    if (view === 'pred') {
      const d = await api('/api/performance/predictions?top_n=20');
      const sc = d.segment_counts||{}, rs = d.role_summary||{}, sd = d.score_distributions||{}, ins = d.auto_insights||[];
      const hp = d.top_high_performers||[], br = d.top_burnout_risk||[], cr = d.critical_employees||[];
      const meth = d.score_methodology||{};
      const sBar = (v, col) => `<div style="display:flex;align-items:center;gap:8px"><div style="flex:1;height:5px;border-radius:3px;background:var(--raised);overflow:hidden"><div style="height:100%;width:${v}%;background:${col};border-radius:3px"></div></div><span style="font-family:var(--mono);font-size:10px;color:var(--muted);min-width:24px">${v}</span></div>`;
      const tagFor = t => { const m={'⭐ High Performer':'green','🔥 Burnout Risk':'amber','⚡ High Productivity':'blue','⚠ Critical — Act Now':'red','✓ Stable':'dim'}; return badge(t, m[t]||'dim'); };
      set('perfContent', `
        <div class="grid-5">
          <div class="stat"><div class="stat-label">Employees</div><div class="stat-value blue">${sc.total_employees}</div></div>
          <div class="stat"><div class="stat-label">High Performers</div><div class="stat-value green">${sc.high_performers}</div><div class="stat-sub">${(sc.high_performers/sc.total_employees*100).toFixed(1)}% of workforce</div></div>
          <div class="stat"><div class="stat-label">Burnout Risk</div><div class="stat-value red">${sc.burnout_risk}</div><div class="stat-sub">score ≥ 60</div></div>
          <div class="stat"><div class="stat-label">High Productivity</div><div class="stat-value purple">${sc.high_productivity}</div><div class="stat-sub">score ≥ 70</div></div>
          <div class="stat"><div class="stat-label">Critical Cases</div><div class="stat-value amber">${sc.critical_cases}</div><div class="stat-sub">high burn + low prod</div></div>
        </div>
        <div class="card"><div class="card-title">Auto Insights</div>${ins.map(i=>`<div class="alert alert-green" style="margin-bottom:6px"><span>◈</span><div class="body">${i}</div></div>`).join('')}</div>
        <div class="grid-3">
          <div class="card"><div class="card-title">Productivity Distribution</div><div class="chart-box" style="height:160px"><canvas id="pfProdDist"></canvas></div></div>
          <div class="card"><div class="card-title">Burnout Distribution</div><div class="chart-box" style="height:160px"><canvas id="pfBurnDist"></canvas></div></div>
          <div class="card"><div class="card-title">High Performer Distribution</div><div class="chart-box" style="height:160px"><canvas id="pfHpDist"></canvas></div></div>
        </div>
        <div class="card">
          <div class="card-title">Scores by Job Role</div>
          <div class="tbl-wrap"><table>
            <thead><tr><th>Role</th><th>Avg Productivity</th><th>Avg Burnout</th><th>Avg HP Score</th><th>Burnout %</th><th>N</th></tr></thead>
            <tbody>${Object.entries(rs).map(([role,v])=>`<tr>
              <td class="val" style="font-size:11px">${role}</td>
              <td style="min-width:120px">${sBar(v.avg_productivity,'var(--green)')}</td>
              <td style="min-width:120px">${sBar(v.avg_burnout,v.avg_burnout>=50?'var(--red)':'var(--amber)')}</td>
              <td style="min-width:120px">${sBar(v.avg_high_performer,'var(--blue)')}</td>
              <td class="${v.burnout_risk_pct>=20?'warn':'val'}">${v.burnout_risk_pct}%</td>
              <td class="val">${v.count}</td>
            </tr>`).join('')}</tbody>
          </table></div>
          <div class="chart-box" style="height:200px;margin-top:16px"><canvas id="pfRoleChart"></canvas></div>
        </div>
        <div class="card ${cr.length?'':''}` + (cr.length?'" style="border-color:rgba(248,113,113,.3)':'"') + `>
          <div class="card-title" ${cr.length?'style="color:var(--red)"':''}>⭐ High Performers (top 15)</div>
          <div class="tbl-wrap"><table>
            <thead><tr><th>Employee</th><th>Role</th><th>HP Score</th><th>Productivity</th><th>Burnout Risk</th><th>Tags</th></tr></thead>
            <tbody>${hp.slice(0,15).map(e=>`<tr>
              <td class="val">${e.emp_id}</td><td style="font-size:11px">${e.job_role}</td>
              <td class="hi">${e.high_performer_score}</td>
              <td class="val">${e.productivity_score}</td>
              <td class="${e.burnout_score>=60?'warn':'val'}">${e.burnout_score}</td>
              <td>${e.tags.map(tagFor).join(' ')}</td>
            </tr>`).join('')}</tbody>
          </table></div>
        </div>
        ${br.length?`<div class="card" style="border-color:rgba(245,158,11,.25)">
          <div class="card-title" style="color:var(--amber)">🔥 Burnout Risk Employees (top 15)</div>
          <div class="tbl-wrap"><table>
            <thead><tr><th>Employee</th><th>Role</th><th>Burnout</th><th>Productivity</th><th>HP Score</th><th>Tags</th></tr></thead>
            <tbody>${br.slice(0,15).map(e=>`<tr>
              <td class="val">${e.emp_id}</td><td style="font-size:11px">${e.job_role}</td>
              <td class="warn">${e.burnout_score}</td>
              <td class="val">${e.productivity_score}</td>
              <td class="val">${e.high_performer_score}</td>
              <td>${e.tags.map(tagFor).join(' ')}</td>
            </tr>`).join('')}</tbody>
          </table></div>
        </div>`:''}
        ${cr.length?`<div class="card" style="border-color:rgba(248,113,113,.35)">
          <div class="card-title" style="color:var(--red)">⚠ Critical Cases — Immediate Attention</div>
          <div class="tbl-wrap"><table>
            <thead><tr><th>Employee</th><th>Role</th><th>Burnout</th><th>Productivity</th></tr></thead>
            <tbody>${cr.map(e=>`<tr><td class="val">${e.emp_id}</td><td style="font-size:11px">${e.job_role}</td><td class="danger">${e.burnout_score}</td><td class="danger">${e.productivity_score}</td></tr>`).join('')}</tbody>
          </table></div>
        </div>`:''}
        <div class="card"><div class="card-title">Score Methodology</div>
          ${Object.entries(meth).map(([k,v])=>`<div style="display:flex;gap:10px;padding:8px 0;border-bottom:1px solid var(--border);font-size:12px"><span style="color:var(--green);font-family:var(--mono);min-width:160px">${k.replace(/_/g,' ')}</span><span style="color:var(--muted)">${v}</span></div>`).join('')}
        </div>`);
      setTimeout(()=>{
        const mkDist=(id,dist,col)=>{const L=Object.keys(dist),V=Object.values(dist);mkChart(id,'bar',L,[{data:V,backgroundColor:col,borderRadius:3}],{scales:{x:{ticks:{color:'#6b7280',maxRotation:45,font:{size:9,family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},y:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});};
        mkDist('pfProdDist',sd.productivity||{},'rgba(0,229,160,.6)');
        mkDist('pfBurnDist',sd.burnout||{},'rgba(248,113,113,.6)');
        mkDist('pfHpDist',sd.high_performer||{},'rgba(56,189,248,.6)');
        const roles=Object.keys(rs);
        mkChart('pfRoleChart','bar',roles,[
          {label:'Productivity',data:roles.map(r=>rs[r].avg_productivity),backgroundColor:'rgba(0,229,160,.6)',borderRadius:3},
          {label:'Burnout Risk',data:roles.map(r=>rs[r].avg_burnout),backgroundColor:'rgba(248,113,113,.55)',borderRadius:3},
          {label:'HP Score',data:roles.map(r=>rs[r].avg_high_performer),backgroundColor:'rgba(56,189,248,.55)',borderRadius:3},
        ],{plugins:{legend:{display:true,position:'top',labels:{color:'#9ca3af',boxWidth:10,font:{family:"'IBM Plex Mono'",size:10}}}},scales:{x:{ticks:{color:'#6b7280',font:{size:9,family:"'IBM Plex Mono'"},maxRotation:30},grid:{color:'rgba(255,255,255,0.04)'}},y:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
      },60);

    } else if (view === 'eda') {
      const d = await api('/api/performance/eda');
      const rolePref = d.role_based_performance||{}, dist = d.target_distribution||{};
      set('perfContent', `
        <div class="grid-3">
          <div class="stat"><div class="stat-label">Records</div><div class="stat-value blue">${d.dataset_info?.rows}</div></div>
          <div class="stat"><div class="stat-label">Features</div><div class="stat-value green">${d.dataset_info?.columns?.length}</div></div>
        </div>
        <div class="grid-2">
          <div class="card"><div class="card-title">Performance by Job Role</div>
            <div class="tbl-wrap"><table><thead><tr><th>Role</th><th>Avg Rating</th><th>Count</th><th>Std Dev</th></tr></thead>
            <tbody>${Object.entries(rolePref).map(([role,s])=>`<tr><td class="val">${role}</td><td class="hi">${s.mean}</td><td class="val">${s.count}</td><td class="val">${s.std}</td></tr>`).join('')}</tbody></table></div>
          </div>
          <div class="card"><div class="card-title">Rating Distribution</div><div class="chart-box" style="height:200px"><canvas id="pfEdaDist"></canvas></div></div>
        </div>`);
      setTimeout(()=>{mkChart('pfEdaDist','bar',Object.keys(dist).map(k=>`Rating ${k}`),[{data:Object.values(dist),backgroundColor:['rgba(56,189,248,.7)','rgba(0,229,160,.7)','rgba(167,139,250,.7)'],borderRadius:4}]);},50);

    } else if (view === 'corr') {
      const d = await api('/api/performance/correlation');
      const pos = d.top_positive_correlates||{}, neg = d.top_negative_correlates||{};
      const all = [...Object.entries(pos).map(([k,v])=>({k,v,dir:'pos'})),...Object.entries(neg).map(([k,v])=>({k,v,dir:'neg'}))].sort((a,b)=>Math.abs(b.v)-Math.abs(a.v));
      set('perfContent', `<div class="card"><div class="card-title">Correlation with PerformanceRating</div>
        <div class="alert alert-blue" style="margin-bottom:16px"><span>◎</span><div class="body">Green = positive. Red = negative. Longer bar = stronger relationship.</div></div>
        ${all.map(c=>`<div class="bar-row"><div class="bar-label">${c.k}</div><div class="bar-track"><div class="bar-fill ${c.dir==='pos'?'green':'red'}" style="width:${Math.abs(c.v)*100}%"></div></div><div class="bar-val" style="color:${c.dir==='pos'?'var(--green)':'var(--red)'}">${c.v}</div></div>`).join('')}
      </div>`);

    } else if (view === 'train') {
      const model = $('perfModel').value, split = $('perfSplit').value;
      const d = await api(`/api/performance/train?model_type=${model}&test_size=${split}`);
      const m = d.metrics, fi = Object.entries(d.feature_importance||{}).slice(0,8), samp = d.sample_predictions||[];
      set('perfContent', `
        <div class="alert alert-green"><span>▲</span><div class="body">${d.interpretation?.accuracy} Top driver: <strong>${d.interpretation?.top_driver}</strong></div></div>
        <div class="grid-4">
          <div class="stat"><div class="stat-label">Accuracy</div><div class="stat-value green">${(m.accuracy*100).toFixed(1)}%</div></div>
          <div class="stat"><div class="stat-label">Precision</div><div class="stat-value blue">${(m.precision*100).toFixed(1)}%</div></div>
          <div class="stat"><div class="stat-label">Recall</div><div class="stat-value purple">${(m.recall*100).toFixed(1)}%</div></div>
          <div class="stat"><div class="stat-label">F1 Score</div><div class="stat-value amber">${(m.f1_score*100).toFixed(1)}%</div></div>
        </div>
        <div class="grid-2">
          <div class="card"><div class="card-title">Feature Importance</div>${fi.map(([n,v])=>`<div class="bar-row"><div class="bar-label">${n}</div><div class="bar-track"><div class="bar-fill blue" style="width:${v/fi[0][1]*100}%"></div></div><div class="bar-val">${v}</div></div>`).join('')}</div>
          <div class="card"><div class="card-title">Sample Predictions</div>
            <div class="tbl-wrap"><table><thead><tr><th>#</th><th>Actual</th><th>Predicted</th><th>✓</th></tr></thead>
            <tbody>${samp.map((p,i)=>`<tr><td>${i+1}</td><td class="val">${p.actual}</td><td class="${p.correct?'hi':'danger'}">${p.predicted}</td><td style="color:${p.correct?'var(--green)':'var(--red)'}">${p.correct?'✓':'✗'}</td></tr>`).join('')}</tbody></table></div>
          </div>
        </div>`);
    }
  } catch(e) { oops('perfContent', e); }
}

File: routers/performance.py
"""
Employee Performance Router
============================
EDA, correlation analysis, and ML model (Random Forest)
to predict PerformanceRating.

Dataset: clinic_performance.csv
"""

import pandas as pd
import numpy as np
from fastapi import APIRouter, Query, Body
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report
from sklearn.preprocessing import LabelEncoder
import warnings
warnings.filterwarnings("ignore")

from .utils import load_csv, df_info, corr_matrix

router = APIRouter()
DEFAULT_FILE = "clinic_performance.csv"
TARGET = "PerformanceRating"

# Columns to drop (ID-like or irrelevant)
DROP_COLS = ["EmpNumber"]


def _preprocess(df: pd.DataFrame) -> tuple[pd.DataFrame, dict]:
    """Encode categoricals, drop irrelevant cols, return processed df + encoding map."""
    df = df.copy()
    df.drop(columns=[c for c in DROP_COLS if c in df.columns], inplace=True)

    encoding_log = {}
    le = LabelEncoder()
    for col in df.select_dtypes(include="object").columns:
        df[col] = le.fit_transform(df[col].astype(str))
        encoding_log[col] = "LabelEncoded"

    # Drop cols with >50% missing
    threshold = len(df) * 0.5
    before = list(df.columns)
    df.dropna(axis=1, thresh=int(threshold), inplace=True)
    dropped = [c for c in before if c not in df.columns]
    if dropped:
        encoding_log["dropped_high_missing"] = dropped

    df.fillna(df.median(numeric_only=True), inplace=True)
    return df, encoding_log


def _train(df: pd.DataFrame, model_type: str, test_size: float):
    """Train chosen model and return metrics + feature importance."""
    if TARGET not in df.columns:
        raise ValueError(f"Target column '{TARGET}' not found.")

    X = df.drop(columns=[TARGET])
    y = df[TARGET]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=test_size, random_state=42, stratify=y
    )

    if model_type == "random_forest":
        model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
    elif model_type == "svm":
        model = SVC(kernel="rbf", probability=True, random_state=42)
    elif model_type == "mlp":
        model = MLPClassifier(hidden_layer_sizes=(64, 32), max_iter=300, random_state=42)
    else:
        model = RandomForestClassifier(n_estimators=100, random_state=42)

    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    metrics = {
        "accuracy": round(accuracy_score(y_test, y_pred), 4),
        "precision": round(precision_score(y_test, y_pred, average="weighted", zero_division=0), 4),
        "recall": round(recall_score(y_test, y_pred, average="weighted", zero_division=0), 4),
        "f1_score": round(f1_score(y_test, y_pred, average="weighted", zero_division=0), 4),
        "train_size": len(X_train),
        "test_size": len(X_test),
        "classes": sorted(y.unique().tolist()),
    }

    report = classification_report(y_test, y_pred, output_dict=True, zero_division=0)
    per_class = {str(k): {m: round(v, 4) for m, v in v2.items()}
                 for k, v2 in report.items() if k not in ("accuracy", "macro avg", "weighted avg")}

    feature_importance = {}
    if hasattr(model, "feature_importances_"):
        fi = dict(zip(X.columns, model.feature_importances_))
        feature_importance = {
            k: round(float(v), 4)
            for k, v in sorted(fi.items(), key=lambda x: x[1], reverse=True)
        }

    # Sample predictions (first 10 test rows)
    sample = pd.DataFrame({
        "actual": y_test.values[:10],
        "predicted": y_pred[:10],
        "correct": (y_test.values[:10] == y_pred[:10]),
    }).to_dict(orient="records")

    return metrics, per_class, feature_importance, sample, model, X.columns.tolist()


@router.get("/eda")
def eda(filename: str = Query(DEFAULT_FILE)):
    """Exploratory Data Analysis: overview, missing values, distributions, role-based analysis."""
    df = load_csv(filename)

    # Role-based performance
    role_col = [c for c in df.columns if "role" in c.lower() or "jobrole" in c.lower().replace(" ", "")]
    role_analysis = {}
    if role_col and TARGET in df.columns:
        role_analysis = df.groupby(role_col[0])[TARGET].agg(["mean", "count", "std"]).round(3).to_dict(orient="index")
        role_analysis = {str(k): {m: round(float(v), 3) if not pd.isna(v) else None for m, v in vals.items()}
                         for k, vals in role_analysis.items()}

    # Column-level summary
    summary = {}
    for col in df.columns:
        col_data = df[col].dropna()
        if df[col].dtype == object:
            summary[col] = {"type": "categorical", "unique": int(col_data.nunique()),
                            "top_values": col_data.value_counts().head(5).to_dict()}
        else:
            summary[col] = {"type": "numeric", "min": round(float(col_data.min()), 3),
                            "max": round(float(col_data.max()), 3),
                            "mean": round(float(col_data.mean()), 3),
                            "std": round(float(col_data.std()), 3)}

    return {
        "dataset_info": df_info(df),
        "column_summary": summary,
        "role_based_performance": role_analysis,
        "target_distribution": df[TARGET].value_counts().to_dict() if TARGET in df.columns else {},
    }


@router.get("/correlation")
def correlation(filename: str = Query(DEFAULT_FILE)):
    """Correlation matrix and top correlates with PerformanceRating."""
    df = load_csv(filename)
    df2, _ = _preprocess(df)
    corr = corr_matrix(df2, target=TARGET)

    top_pos = {}
    top_neg = {}
    if TARGET in df2.columns:
        target_corr = pd.DataFrame(df2.corr()[TARGET]).drop(TARGET, errors="ignore")
        top_pos = target_corr[target_corr[TARGET] > 0].sort_values(TARGET, ascending=False).head(5)[TARGET].round(4).to_dict()
        top_neg = target_corr[target_corr[TARGET] < 0].sort_values(TARGET).head(5)[TARGET].round(4).to_dict()

    return {
        "correlation_matrix": corr,
        "top_positive_correlates": top_pos,
        "top_negative_correlates": top_neg,
    }


@router.get("/train")
def train_model(
    filename: str = Query(DEFAULT_FILE),
    model_type: str = Query("random_forest", description="random_forest | svm | mlp"),
    test_size: float = Query(0.2, ge=0.1, le=0.4, description="Test split fraction"),
):
    """Train an ML model to predict PerformanceRating. Returns metrics, feature importance, sample predictions."""
    df = load_csv(filename)
    df2, encoding_log = _preprocess(df)
    metrics, per_class, feature_importance, sample, _, features = _train(df2, model_type, test_size)

    return {
        "model": model_type,
        "target": TARGET,
        "features_used": features,
        "preprocessing": encoding_log,
        "metrics": metrics,
        "per_class_metrics": per_class,
        "feature_importance": feature_importance,
        "sample_predictions": sample,
        "interpretation": {
            "accuracy": f"The model correctly predicts performance {metrics['accuracy']*100:.1f}% of the time.",
            "top_driver": next(iter(feature_importance)) if feature_importance else "N/A",
            "recommendation": (
                "Focus on the top feature importance drivers to improve employee performance. "
                "High salary hike % and job involvement strongly predict higher ratings."
            ),
        },
    }


# ── Score formulas ───────────────────────────────────────────────────────────
def _add_scores(df_enc: pd.DataFrame, df_orig: pd.DataFrame) -> pd.DataFrame:
    """Compute three composite scores using original (unencoded) column values."""
    d = df_enc.copy()

    # Productivity (0-100)
    d["productivity_score"] = (
        ((df_orig["PerformanceRating"] - 2) / 2) * 40 +
        ((df_orig["EmpJobInvolvement"]  - 1) / 3) * 25 +
        ((df_orig["EmpJobSatisfaction"] - 1) / 3) * 20 +
        ((df_orig["EmpEnvironmentSatisfaction"] - 1) / 3) * 15
    ).clip(0, 100).round(1).values

    # Burnout risk (0-100)
    promo_max = max(df_orig["YearsSinceLastPromotion"].max(), 1)
    dist_max  = max(df_orig["DistanceFromHomeKm"].max(), 1)
    ot        = (df_orig["OverTime"] == "Yes").astype(float)
    d["burnout_score"] = (
        ot * 30 +
        ((4 - df_orig["EmpWorkLifeBalance"]) / 3) * 25 +
        ((4 - df_orig["EmpJobSatisfaction"]) / 3) * 20 +
        (df_orig["YearsSinceLastPromotion"] / promo_max) * 15 +
        (df_orig["DistanceFromHomeKm"] / dist_max) * 10
    ).clip(0, 100).round(1).values

    # High performer score (0-100)
    hike_max = max(df_orig["EmpLastSalaryHikePercent"].max(), 1)
    d["high_performer_score"] = (
        ((df_orig["PerformanceRating"] - 2) / 2) * 50 +
        (df_orig["EmpLastSalaryHikePercent"] / hike_max) * 30 +
        ((df_orig["EmpJobInvolvement"] - 1) / 3) * 20
    ).clip(0, 100).round(1).values

    return d


def _label(hp: float, burn: float, prod: float) -> dict:
    tags = []
    if hp >= 70:   tags.append("⭐ High Performer")
    if burn >= 60: tags.append("🔥 Burnout Risk")
    if prod >= 70: tags.append("⚡ High Productivity")
    if burn >= 60 and prod < 40: tags.append("⚠ Critical — Act Now")
    if not tags:   tags.append("✓ Stable")
    return {
        "high_performer_score": hp,
        "burnout_score": burn,
        "productivity_score": prod,
        "tags": tags,
        "burnout_level": "High" if burn >= 60 else ("Medium" if burn >= 35 else "Low"),
        "performance_tier": "Top" if hp >= 70 else ("Mid" if hp >= 40 else "Low"),
    }


@router.get("/predictions")
def predictions(
    filename: str = Query(DEFAULT_FILE),
    top_n: int = Query(20, ge=5, le=100, description="Employees to return per list"),
):
    """
    Per-employee predictions:
    - High Performer score (0-100)
    - Burnout Risk score (0-100)
    - Productivity score (0-100)
    - Segment labels and tier per employee
    - Aggregate distributions and role-level summaries
    - Auto-generated insights
    """
    df = load_csv(filename)
    df_raw, _ = _preprocess(df.copy())

    scored = _add_scores(df_raw, df)
    scored["EmpNumber"] = df["EmpNumber"].values
    scored["EmpJobRole"] = df["EmpJobRole"].values

    # ── Per-employee table ────────────────────────────────────────
    records = []
    for _, row in scored.iterrows():
        hp  = float(row["high_performer_score"])
        burn = float(row["burnout_score"])
        prod = float(row["productivity_score"])
        info = _label(hp, burn, prod)
        records.append({
            "emp_id":   str(row["EmpNumber"]),
            "job_role": str(row["EmpJobRole"]),
            **info,
        })

    records.sort(key=lambda x: x["high_performer_score"], reverse=True)

    # ── Segment counts ────────────────────────────────────────────
    high_performers = [r for r in records if r["high_performer_score"] >= 70]
    burnout_risk    = [r for r in records if r["burnout_score"] >= 60]
    high_prod       = [r for r in records if r["productivity_score"] >= 70]
    critical        = [r for r in records if r["burnout_score"] >= 60 and r["productivity_score"] < 40]

    # ── Role-level summary ────────────────────────────────────────
    role_summary = {}
    for role, grp in scored.groupby("EmpJobRole"):
        role_summary[str(role)] = {
            "count": int(len(grp)),
            "avg_productivity": float(grp["productivity_score"].mean().round(1)),
            "avg_burnout":      float(grp["burnout_score"].mean().round(1)),
            "avg_high_performer": float(grp["high_performer_score"].mean().round(1)),
            "high_performers_pct": round(float((grp["high_performer_score"] >= 70).mean() * 100), 1),
            "burnout_risk_pct":   round(float((grp["burnout_score"] >= 60).mean() * 100), 1),
        }

    # ── Score distributions (histogram buckets) ──────────────────
    def hist(series, bins):
        counts, edges = np.histogram(series, bins=bins)
        labels = [f"{int(edges[i])}–{int(edges[i+1])}" for i in range(len(edges)-1)]
        return {labels[i]: int(counts[i]) for i in range(len(labels))}

    bins10 = list(range(0, 110, 10))
    prod_dist  = hist(scored["productivity_score"],    bins10)
    burn_dist  = hist(scored["burnout_score"],         bins10)
    hp_dist    = hist(scored["high_performer_score"],  bins10)

    # ── Auto insights ─────────────────────────────────────────────
    total = len(records)
    insights = [
        f"{len(high_performers)} employees ({len(high_performers)/total*100:.1f}%) qualify as High Performers (score ≥ 70).",
        f"{len(burnout_risk)} employees ({len(burnout_risk)/total*100:.1f}%) are at Burnout Risk (score ≥ 60). Immediate attention recommended.",
        f"{len(critical)} employees show both high burnout risk AND low productivity — critical intervention needed.",
        f"Average productivity score across all roles: {scored['productivity_score'].mean():.1f}/100.",
        f"Average burnout risk score: {scored['burnout_score'].mean():.1f}/100.",
    ]
    top_burn_role = max(role_summary, key=lambda r: role_summary[r]["burnout_risk_pct"])
    top_prod_role = max(role_summary, key=lambda r: role_summary[r]["avg_productivity"])
    insights.append(f"Highest burnout risk role: {top_burn_role} ({role_summary[top_burn_role]['burnout_risk_pct']}% at risk).")
    insights.append(f"Highest average productivity role: {top_prod_role} ({role_summary[top_prod_role]['avg_productivity']}/100).")

    return {
        "dataset_info": df_info(df),
        "segment_counts": {
            "total_employees": total,
            "high_performers": len(high_performers),
            "burnout_risk": len(burnout_risk),
            "high_productivity": len(high_prod),
            "critical_cases": len(critical),
        },
        "top_high_performers":  records[:top_n],
        "top_burnout_risk":     sorted(burnout_risk,  key=lambda x: x["burnout_score"],      reverse=True)[:top_n],
        "top_productive":       sorted(high_prod,      key=lambda x: x["productivity_score"], reverse=True)[:top_n],
        "critical_employees":   critical,
        "role_summary":         role_summary,
        "score_distributions": {
            "productivity":     prod_dist,
            "burnout":          burn_dist,
            "high_performer":   hp_dist,
        },
        "auto_insights": insights,
        "score_methodology": {
            "productivity_score": "40% PerformanceRating + 25% JobInvolvement + 20% JobSatisfaction + 15% EnvironmentSatisfaction",
            "burnout_score":      "30% OverTime + 25% low WorkLifeBalance + 20% low JobSatisfaction + 15% YearsSinceLastPromotion + 10% DistanceFromHome",
            "high_performer_score": "50% PerformanceRating + 30% SalaryHike% + 20% JobInvolvement",
        },
    }
